README.md
# mongoose_serial_server
mongoose motor driver and low level ROS api package
    
## input topic
      name                                  type        
    /cmd_vel                         geometry_msgs/Twist
      
## output topic
       name                                 type                 frate
    /mongoose_serial_server/Odom            nav_msgs/Odometry            50hz
    /mongoose_serial_server/Pose2D          geometry_msgs/Pose2D         50hz
    /mongoose_serial_server/Power           std_msgs/Float64             50hz
    /mongoose_serial_server/StatusFlag      std_msgs/Int32               50hz
    /mongoose_serial_server/Twist           geometry_msgs/Twist          50hz
    /tf                                     odom-->base_footprint        50hz
    /tf_static                              base_footprint-->base_link   100hz
    
## input param   
       name                            default
    port                             /dev/ttyUSB0
    baud                               115200
    wheel_separation                    0.47
    wheel_radius                        0.1016
    debug_flag                          false
    max_speed                           0.6
    cmd_topic                           cmd_vel

## Usage:
### download to ros workspace
```
cd ~/catkin_ws/src
git clone https://github.com/lwc3258/MongooseSerialServer.git
cd ..
catkin_make
```
### Quickstart
```
roslaunch mongoose_serial_server mongoose_serial_server.launch
```



CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package mongoose_serial_server
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.1 (2017-09-9)
-------------------
* Update code for control the AGV run in a 6 x 2 square.
* Release the code as 1.0.1 version.



src/StatusPublisher.h
#ifndef STATUSPUBLISHER_H
#define STATUSPUBLISHER_H

#include "ros/ros.h"
#include <boost/thread.hpp>
#include <boost/assign/list_of.hpp>
#include <algorithm>
#include "geometry_msgs/Pose2D.h"
#include "geometry_msgs/Twist.h"
#include "geometry_msgs/Vector3.h"
#include "nav_msgs/Odometry.h"
#include "std_msgs/Int32.h"
#include "std_msgs/Float64.h"
#include "tf/transform_broadcaster.h"
#include <sensor_msgs/point_cloud2_iterator.h>
#include <sensor_msgs/PointField.h>


#include <list>
#include <time.h>
// #include <chrono>

#define PI 3.14159265

namespace xqserial_server
{
typedef struct {
    int status;//小车状态，0表示未初始化，1表示正常，-1表示error
    float power;//电源电压【9 13】v
    float theta;//方位角，【0 360）°

    int encoder_ppr;//车轮1转对应的编码器个数
    int encoder_delta_r;//右轮编码器增量， 个为单位
    int encoder_delta_l;//左轮编码器增量， 个为单位
    int encoder_delta_car;//两车轮中心位移，个为单位

    int omga_r;//右轮转速 个每秒
    int omga_l;//左轮转速 个每秒
    int omga_FR;
    int omga_FL;
    int omga_BL;
    int omga_BR;

    float velocity_x;
    float velocity_y;
    float angular_vel;


    float distance1;//第一个超声模块距离值 单位cm
    float distance2;//第二个超声模块距离值 单位cm
    float distance3;//第三个超声模块距离值 单位cm
    float distance4;//第四个超声模块距离值 单位cm

    float IMU[9];//mpu6050 9轴数据
    float gyro;
    float m_gyro_angular_vel;
    float m_gyro_acc_x;

    unsigned int m_obstacle_state;
    unsigned int time_stamp;//时间戳
    

}UPLOAD_STATUS;

class StatusPublisher
{

public:
    StatusPublisher();

    StatusPublisher(double separation,double radius);

    void Refresh();

    void Update(const char *data, unsigned int len);

    void encoder_IMU_parse(std::vector<unsigned char> data);

    bool crc_check(std::vector<unsigned char> data);

    double get_wheel_separation();

    double get_wheel_radius();

    int get_wheel_ppr();

    int get_status();

    void get_wheel_speed(double speed[2]);

    geometry_msgs::Pose2D get_CarPos2D();

    geometry_msgs::Twist get_CarTwist();

    std_msgs::Float64 get_power();

    nav_msgs::Odometry get_odom();

    UPLOAD_STATUS car_status;
    
    //data type used for compute the sensor data average value.
    std::list<float> ml_position_x;
    std::list<float> ml_position_y;
    std::list<float> ml_gyro_theta;
    std::list<float> ml_gyro_angular_vel;
    std::list<float> ml_encoder_vx;
    std::list<float> ml_encoder_vy;

    float mf_average_position_x;
    float mf_average_position_y;
    float mf_average_gyro_theta;
    float mf_average_gyro_angular_vel;
    float mf_average_encoder_vx;
    float mf_average_encoder_vy;

    //obstacle state from stm32.
    int g_obstacle_state;

private:

    //Wheel separation, wrt the midpoint of the wheel width: meters
    double wheel_separation;

    //Wheel radius (assuming it's the same for the left and right wheels):meters
    double wheel_radius;

    unsigned long package_last_time_stamp;

    geometry_msgs::Pose2D CarPos2D;//小车开始启动原点坐标系
    geometry_msgs::Twist  CarTwist;//小车自身坐标系
    std_msgs::Float64 CarPower;// 小车电池信息
    nav_msgs::Odometry CarOdom;// 小车位置和速度信息

    ros::NodeHandle mNH;
    ros::Publisher mPose2DPub;
    ros::Publisher mTwistPub;
    ros::Publisher mStatusFlagPub;
    ros::Publisher mPowerPub;
    ros::Publisher mOdomPub;
    ros::Publisher pub_barpoint_cloud_;
    ros::Publisher pub_clearpoint_cloud_;

    bool mbUpdated;

    boost::mutex mMutex;
};

} //namespace xqserial_server


#endif // STATUSPUBLISHER_H


src/DiffDriverController.h
#ifndef DIFFDRIVERCONTROLLER_H
#define DIFFDRIVERCONTROLLER_H
#include <ros/ros.h>
#include "StatusPublisher.h"
#include "AsyncSerial.h"
#include <std_msgs/Bool.h>

#define INFORMATION_LENGTH 14

namespace xqserial_server
{

class DiffDriverController
{
public:
    DiffDriverController();

    DiffDriverController(double max_speed_,std::string cmd_topic_,StatusPublisher* xq_status_,CallbackAsyncSerial* cmd_serial_);

    void run();

    void sendcmd(const geometry_msgs::Twist& command);

    void sendcmdTest(const geometry_msgs::Twist &command);

    void imuCalibration(const std_msgs::Bool& calFlag);

    void setStatusPtr(StatusPublisher& status);

    void updateMoveFlag(const std_msgs::Bool& moveFlag);

    void updateBarDetectFlag(const std_msgs::Bool& DetectFlag);

private:
    double max_wheelspeed;//单位为转每秒,只能为正数

    std::string cmd_topic;

    StatusPublisher* xq_status;

    CallbackAsyncSerial* cmd_serial;

    boost::mutex mMutex;
    
    bool MoveFlag;
};

}
#endif // DIFFDRIVERCONTROLLER_H


src/DiffDriverController.cpp
#include "DiffDriverController.h"
#include <time.h>
#include <stdint.h>

namespace xqserial_server
{

DiffDriverController::DiffDriverController()
{
    max_wheelspeed = 2.0;
    cmd_topic = "cmd_vel";
    xq_status = new StatusPublisher();
    cmd_serial = NULL;
    MoveFlag = true;
}

DiffDriverController::DiffDriverController(double max_speed_, std::string cmd_topic_, StatusPublisher *xq_status_, CallbackAsyncSerial *cmd_serial_)
{
    MoveFlag = true;
    max_wheelspeed = max_speed_;
    cmd_topic = cmd_topic_;
    xq_status = xq_status_;
    cmd_serial = cmd_serial_;
}

void DiffDriverController::run()
{
    ros::NodeHandle nodeHandler;
    ros::Subscriber sub = nodeHandler.subscribe(cmd_topic, 1, &DiffDriverController::sendcmd, this);
    ros::Subscriber sub2 = nodeHandler.subscribe("/imu_cal", 1, &DiffDriverController::imuCalibration, this);
    ros::Subscriber sub3 = nodeHandler.subscribe("/globalMoveFlag", 1, &DiffDriverController::updateMoveFlag, this);
    ros::Subscriber sub4 = nodeHandler.subscribe("/barDetectFlag", 1, &DiffDriverController::updateBarDetectFlag, this);
    ros::spin();
}

void DiffDriverController::updateMoveFlag(const std_msgs::Bool &moveFlag)
{
    boost::mutex::scoped_lock lock(mMutex);
    MoveFlag = moveFlag.data;
}

void DiffDriverController::imuCalibration(const std_msgs::Bool &calFlag)
{
    if (calFlag.data)
    {
        //下发底层ｉｍｕ标定命令
        char cmd_str[5] = {0xcd, 0xeb, 0xd7, 0x01, 0x43};
        if (NULL != cmd_serial)
        {
            cmd_serial->write(cmd_str, 5);
        }
    }
}

void DiffDriverController::updateBarDetectFlag(const std_msgs::Bool &DetectFlag)
{
    if (DetectFlag.data)
    {
        //下发底层红外开启命令
        char cmd_str[6] = {0xcd, 0xeb, 0xd7, 0x02, 0x44, 0x01};
        if (NULL != cmd_serial)
        {
            cmd_serial->write(cmd_str, 6);
        }
    }
    else
    {
        //下发底层红外禁用命令
        char cmd_str[6] = {0xcd, 0xeb, 0xd7, 0x02, 0x44, 0x00};
        if (NULL != cmd_serial)
        {
            cmd_serial->write(cmd_str, 6);
        }
    }
}

void DiffDriverController::sendcmd(const geometry_msgs::Twist &command)
{
    static time_t t1 = time(NULL), t2;
    int i = 0, wheel_ppr = 1;
    int crc_length = 0;
    int crc = 0;
    bool stop_cmd = false;
    double separation = 0, radius = 0, speed_temp[2];
    int speed_lin_x = 0, speed_lin_y = 0, speed_ang = 0;
    char cmd_str[14] = {0xFF, 0x03, 0x09, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xF5, 0xED};
    char cmd_str1[14] = {0xFF, 0x03, 0x09, 0x00, 0x1E, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xEB, 0xED};
    
    char stop_str[5] = {0xFF, 0x04, 0x00, 0xFB, 0xED};

    // if(xq_status->get_status()==0) return;//底层还在初始化

    separation = xq_status->get_wheel_separation();
    radius = xq_status->get_wheel_radius();
    wheel_ppr = xq_status->get_wheel_ppr();

    //坐标系转换，将move_base的坐标系转成小车的坐标系(World and AGV coordinates are the same)。
    speed_lin_x = command.linear.x * 100;
    speed_lin_y = command.linear.y * 100;
    speed_ang = command.angular.z * 100;
    // std::cout << "cmmond angular z:" << command.angular.z << std::endl;
    // std::cout << "speed_ang angular z:" << speed_ang << std::endl;

     //Get theta value.
    float speed[3] = {speed_lin_x, speed_lin_y, speed_ang};

    //put stop and velocity cmd.
    if (speed[0] == 0 && speed[1] == 0 && speed[2] == 0)
    {
        stop_cmd = true;
    }
    else
    {
        // std::cout << "This line:" << __LINE__ << std::endl;

        //put processed value to cmd_str.
        int Vx_state = 0, Vy_state = 0, Wz_state = 0;
        int Vx_tmp, Vy_tmp, Wz_tmp;
        uint32_t speed_vx, speed_vy, speed_wz;

        //vx
        if (speed[0] >= 0)
        {
            Vx_state = 0;
            speed_vx = speed[0];
        }
        else
        {
            Vx_state = 1;
            speed_vx = - speed[0];
        }
       
        Vx_tmp = ( speed_vx & 0x0000ff00) >> 8;
        cmd_str[3] = Vx_tmp;

        Vx_tmp = (speed_vx & 0x000000ff);
        cmd_str[4] = Vx_tmp;
        cmd_str[5] = Vx_state;

        //vy
        if (speed[1] >= 0)
        {
            Vy_state = 0;
            speed_vy = speed[1];
        }
        else
        {
            Vy_state = 1;
            speed_vy = -speed[1];
        }

        Vy_tmp = (speed_vy & 0x0000ff00) >> 8;
        cmd_str[6] = Vy_tmp;

        Vy_tmp = (speed_vy & 0x000000ff);
        cmd_str[7] = Vy_tmp;
        cmd_str[8] = Vy_state;

        //Wz
        if(speed[2] >= 0)
        {
            Wz_state = 0;
            speed_wz = speed[2];
        }
        else
        {
            Wz_state = 1;
            speed_wz = -speed[2];
        }

        Wz_tmp = (speed_wz & 0x0000ff00) >> 8;
        cmd_str[9] = Wz_tmp;

        Wz_tmp = (speed_wz & 0x000000ff);
        cmd_str[10] = Wz_tmp;
        cmd_str[11] = Wz_state;

        crc_length = (INFORMATION_LENGTH - 2);
        for (i = 0; i < crc_length; ++i)
        {
            crc = crc ^ cmd_str[i];
        }
        cmd_str[12] = (char)crc;

        cmd_str[13] = 0xED;
    }

    boost::mutex::scoped_lock lock(mMutex);
    if (!MoveFlag)
    {
        cmd_str[5] = 0x53;
        cmd_str[6] = 0x53;
    }
    if (NULL != cmd_serial)
    {
        if (stop_cmd)
        {
            cmd_serial->write(stop_str, 5);
        }
        else
        {
            cmd_serial->write(cmd_str, 14);
        }
    }
}

}


src/main.cpp

#include "AsyncSerial.h"

#include <iostream>
#include <boost/thread.hpp>

#include <ros/console.h>
#include <ros/ros.h>
#include <ros/package.h>
#include "DiffDriverController.h"
#include "StatusPublisher.h"

using namespace std;

int main(int argc, char **argv)
{

    if (ros::console::set_logger_level(ROSCONSOLE_DEFAULT_NAME, ros::console::levels::Info)) { // Change the level to fit your needs
        ros::console::notifyLoggerLevelsChanged();
     }

    ROS_INFO("welcome to Mongoose serial server !");

    ros::init(argc, argv, "xqserial_server");
    ros::start();

    //获取串口参数
    std::string port;
    ros::param::param<std::string>("~port", port, "/dev/ttyUSB0");
    int baud;
    ros::param::param<int>("~baud", baud, 115200);
    cout << "port:" << port << " baud:" << baud << endl;

    //获取小车机械参数
    double separation = 0, radius = 0;
    bool DebugFlag = false;
    ros::param::param<double>("~wheel_separation", separation, 0.47);
    ros::param::param<double>("~wheel_radius", radius, 0.1016);
    ros::param::param<bool>("~debug_flag", DebugFlag, false);
    xqserial_server::StatusPublisher xq_status(separation, radius);

    //获取小车控制参数
    double max_speed;
    string cmd_topic;
    ros::param::param<double>("~max_speed", max_speed, 2.0);
    ros::param::param<std::string>("~cmd_topic", cmd_topic, "cmd_vel");

    try
    {
        CallbackAsyncSerial serial(port, baud);

        serial.setCallback(boost::bind(&xqserial_server::StatusPublisher::Update, &xq_status, _1, _2));
        xqserial_server::DiffDriverController xq_diffdriver(max_speed, cmd_topic, &xq_status, &serial);
        boost::thread cmd2serialThread(&xqserial_server::DiffDriverController::run, &xq_diffdriver);

        // send test flag
        char debugFlagCmd[] = {0xcd, 0xeb, 0xd7, 0x01, 'T'};
        if (DebugFlag)
        {
            std::cout << "Debug mode Enabled" << std::endl;
            serial.write(debugFlagCmd, 5);
        }

         char cmd_str1[14] = {0xFF, 0x03, 0x09, 0x00, 0x1E, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xEB, 0xED};

         //发布周期为50hz
        ros::Rate r(50); 
        while (ros::ok())
        {
            if (serial.errorStatus() || serial.isOpen() == false)
            {
                cerr << "Error: serial port closed unexpectedly" << endl;
                break;
            }

            //定时发布状态小车状态
            xq_status.Refresh(); 
            r.sleep();
        }

    quit:
        serial.close();
    }
    catch (std::exception &e)
    {
        cerr << "Exception: " << e.what() << endl;
    }

    ros::shutdown();
    return 0;
}
